### ISSUE TITLE
[Feature] adapt fused sigmoid gate for MoE model

### ISSUE THREAD

**Issue by zhyncs** [sgl-project/sglang#2739]
### Checklist

- [ ] 1. If the issue you raised is not a feature but a question, please raise a discussion at https://github.com/sgl-project/sglang/discussions/new/choose Otherwise, it will be closed.
- [ ] 2. Please use English, otherwise it will be closed.

### Motivation

ref https://github.com/NVIDIA/TensorRT-LLM/blob/be1788106245496872d18e702978e59b6bfd50e0/cpp/tensorrt_llm/kernels/mixtureOfExperts/moe_kernels.cu#L232

**Comment by zhaochenyang20** [comment]
@NovTi

**Comment by sitabulaixizawaluduo** [comment]
> ### Checklist
> * [ ]  1. If the issue you raised is not a feature but a question, please raise a discussion at https://github.com/sgl-project/sglang/discussions/new/choose Otherwise, it will be closed.
> * [ ]  2. Please use English, otherwise it will be closed.
> 
> ### Motivation
> ref https://github.com/NVIDIA/TensorRT-LLM/blob/be1788106245496872d18e702978e59b6bfd50e0/cpp/tensorrt_llm/kernels/mixtureOfExperts/moe_kernels.cu#L232
> 
> ### Related resources
> _No response_

LMDeploy should also do the same optimization, which should greatly improve the performance of MoE

**Comment by zixuanzhang226** [comment]
Hello,

I would like to take on this task and share some findings with you.

We ran DeepSeek-R1 on SGLang and verified that it follows the logic of biased_grouped_topk inside the select_experts function in python/sglang/srt/layers/moe/topk.py. Within the biased_grouped_topk function (line 126), we confirmed that scores = gating_output.sigmoid() is already implemented.

Also, I tested the performance of DeepSeek R1 using softmax vs sigmoid, and the tokens/s for both are quite similar. The average tokens/s of using softmax is about 5.32, and the average tokens/s of using sigmoid is about 5.30. Please see the test results below.

Here are two proposed changes:

1. Update the comment on line 166 in python/sglang/srt/layers/moe/topk.py to "# DeepSeek V2/V3/R1 uses biased_grouped_top" for better clarity and to eliminate potential confusion.

2. Modify the grouped_topk function in python/sglang/srt/layers/moe/topk.py on line 90 to support both softmax and sigmoid scoring functions. The proposed pseudo code change:

```
if scoring_func == "softmax":
    scores = torch.softmax(gating_output, dim=-1)
elif scoring_func == "sigmoid":
    scores = gating_output.sigmoid()
else:
    raise ValueError("xxxx")
```

Please let me know your thoughts. Thank you very much. 

<img width="1213" alt="Image" src="https://github.com/user-attachments/assets/f073aba8-ee08-4c84-b0c8-dac973889f40" />

<img width="1209" alt="Image" src="https://github.com/user-attachments/assets/5f9009c6-38ee-4055-aba7-fb7ae54823cb" />

**Comment by zhaochenyang20** [comment]
@FrankLeeeee could you please take a look? Thanks!

**Comment by zixuanzhang226** [comment]
Hello @FrankLeeeee  would you please review it? Thank you very much!

**Comment by zhaochenyang20** [comment]
Thanks! @zixuanzhang226 I will call him

**Comment by FrankLeeeee** [comment]
Hi @zixuanzhang226 , I think your proposal looks good to me, do you want to take this task and contribute a PR?

**Comment by zixuanzhang226** [comment]
Hello @FrankLeeeee , yes, I would like to take this task and I will send out the PR later. Thank you!

**Comment by FrankLeeeee** [comment]
@zixuanzhang226 appreicate it, thanks for your incoming contribution!

**Comment by zhaochenyang20** [comment]
@FrankLeeeee Are there PRs for this?

**Comment by FrankLeeeee** [comment]
> [@FrankLeeeee](https://github.com/FrankLeeeee) Are there PRs for this?

@zixuanzhang226 will put up a PR.

**Comment by NovTi** [comment]
> Hello [@FrankLeeeee](https://github.com/FrankLeeeee) , yes, I would like to take this task and I will send out the PR later. Thank you!

Hi Zixuan, may I ask what are you planning to do on this issue? I have already implemented the fused sigmoid gate in this pr #3191. But the performance improvement is minor, Yineng suggested hold this issue

**Comment by zixuanzhang226** [comment]
Hello @FrankLeeeee , would you please take a look the PR https://github.com/sgl-project/sglang/pull/3680? Appreciate that. 
Additionally, we have one concern: As we previously ran DeepSeek-R1 on SGLang and confirmed that it follows the logic of biased_grouped_topk, we would like to double-check with you to ensure that this is expected behavior—meaning it follows biased_grouped_topk rather than grouped_topk, right?

**Comment by zixuanzhang226** [comment]
Hello @NovTi , I reviewed your PR, and it should have any conflicts with yours~ I just did an improvement on the grouped_topk logic.

**Comment by FrankLeeeee** [comment]
> Hello [@FrankLeeeee](https://github.com/FrankLeeeee) , would you please take a look the PR [#3680](https://github.com/sgl-project/sglang/pull/3680)? Appreciate that. Additionally, we have one concern: As we previously ran DeepSeek-R1 on SGLang and confirmed that it follows the logic of biased_grouped_topk, we would like to double-check with you to ensure that this is expected behavior—meaning it follows biased_grouped_topk rather than grouped_topk, right?

Honestly, I got to check on this. Do you check the DPSK original implementation? @Fridge003 do you have any idea

**Comment by NovTi** [comment]
> Hello [@NovTi](https://github.com/NovTi) , I reviewed your PR, and it should have any conflicts with yours~ I just did an improvement on the grouped_topk logic.

Cooooool

**Comment by Fridge003** [comment]
> > Hello [@FrankLeeeee](https://github.com/FrankLeeeee) , would you please take a look the PR [#3680](https://github.com/sgl-project/sglang/pull/3680)? Appreciate that. Additionally, we have one concern: As we previously ran DeepSeek-R1 on SGLang and confirmed that it follows the logic of biased_grouped_topk, we would like to double-check with you to ensure that this is expected behavior—meaning it follows biased_grouped_topk rather than grouped_topk, right?
> 
> Honestly, I got to check on this. Do you check the DPSK original implementation? [@Fridge003](https://github.com/Fridge003) do you have any idea

LGTM

I only have one question. Is this issue trying to adapt Cuda kernel used in TRT-LLM?
If triton kernel generated by pytorch compile is accepted, then it will be OK

**Comment by zixuanzhang226** [comment]
@FrankLeeeee Yes, I checked the code, and it does use the logic of `correction_bias = self.gate.e_score_correction_bias` (https://github.com/sgl-project/sglang/blob/6b0aeb58fd530f88defd5c7862c74a5c7c1a5dba/python/sglang/srt/models/deepseek_v2.py#L156). By the way, I rebased with main. Thanks!

**Comment by FrankLeeeee** [comment]
> [@FrankLeeeee](https://github.com/FrankLeeeee) Yes, I checked the code, and it does use the logic of `correction_bias = self.gate.e_score_correction_bias` (
> 
> [sglang/python/sglang/srt/models/deepseek_v2.py](https://github.com/sgl-project/sglang/blob/6b0aeb58fd530f88defd5c7862c74a5c7c1a5dba/python/sglang/srt/models/deepseek_v2.py#L156)
> 
> Line 156 in [6b0aeb5](/sgl-project/sglang/commit/6b0aeb58fd530f88defd5c7862c74a5c7c1a5dba)
> 
>  correction_bias=self.gate.e_score_correction_bias, 
> ). By the way, I rebased with main. Thanks!

Awesome!

**Comment by b8zhong** [comment]
Complete in https://github.com/sgl-project/sglang/pull/3680
